packages/ra-core/src/form/useInput.ts

```diff
@@ -67,8 +70,12 @@ const useInput = ({
         ? composeValidators(validate)
         : validate;
 
+    // Fetch the initialValue from the record if available or apply the provided initialValue.
+    // This ensure dynamically added inputs have their value set correctly (ArrayInput for example).
+    // We don't do this for the form level initialValues so that it works as it should in final-form
+    // (ie. field level initialValue override form level initialValues for this field).
     const { input, meta } = useFinalFormField(finalName, {
-        initialValue,
+        initialValue: get(record, source, initialValue),
```

Comment:
If I may, this makes sense to me.

Reply:
Yes, we were actually lucky it worked before XD